File: src/main/java/com/kigya/patterns/java/strategy/entity/Duck.java
package com.kigya.patterns.java.strategy.entity;

import com.kigya.patterns.java.strategy.behaviour.FlyBehaviour;
import com.kigya.patterns.java.strategy.behaviour.QuackBehaviour;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.extern.flogger.Flogger;

import java.util.logging.Level;

@Setter
@Flogger
@NoArgsConstructor(access = AccessLevel.PACKAGE)
public abstract class Duck {

    FlyBehaviour flyBehaviour;
    QuackBehaviour quackBehaviour;

    abstract void display();

    public void performFly() {
        flyBehaviour.fly();
    }

    public void performQuack() {
        quackBehaviour.quack();
    }

    public void swim() {
        log.at(Level.INFO).log("All ducks float, even decoys!", Duck.class.toString());
    }
}

File: src/main/java/com/kigya/patterns/kotlin/strategy/entity/Duck.kt
package com.kigya.patterns.kotlin.strategy.entity

import com.kigya.patterns.java.strategy.entity.Duck
import mu.KotlinLogging

typealias Quack = () -> Unit
typealias Fly = () -> Unit

val logger = KotlinLogging.logger { }

abstract class Duck {

    abstract var quackBehaviour: Quack
    abstract var flyBehaviour: Fly

    abstract fun display()

    fun performQuack() = quackBehaviour.invoke()
    fun performFly() = flyBehaviour.invoke()

    open fun swim() = logger.info {
        "All ducks float, even decoys: " + this::class.simpleName.toString()
    }

}

fun muteQuack() = logger.info { "Silence: " + Duck::performQuack::class.java.toString() }
fun squeak() = logger.info { "Squeak: " + Duck::performQuack::class.java.toString()  }
fun flyNoWay() = logger.info { "I can't fly: " + Duck::performFly::class.java.toString()}
fun flyWithWings() = logger.info { "I'm flying!: " + Duck::performFly::class.java.toString()}